clients/src/main/java/org/apache/kafka/common/serialization/ListDeserializer.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.serialization;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.common.errors.SerializationException;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+
+public class ListDeserializer<Inner> implements Deserializer<List<Inner>> {
+
+    private Deserializer<Inner> inner;
+    private Class<?> listClass;
+    private Integer primitiveSize;
+
+    static private Map<Class<? extends Deserializer>, Integer> fixedLengthDeserializers = mkMap(
+        mkEntry(ShortDeserializer.class, 2),
+        mkEntry(IntegerDeserializer.class, 4),
+        mkEntry(FloatDeserializer.class, 4),
+        mkEntry(LongDeserializer.class, 8),
+        mkEntry(DoubleDeserializer.class, 8),
+        mkEntry(UUIDDeserializer.class, 36)
+    );
+
+    public ListDeserializer() {}
+
+    public <L extends List> ListDeserializer(Class<L> listClass, Deserializer<Inner> innerDeserializer) {
+        this.listClass = listClass;
+        this.inner = innerDeserializer;
+        this.primitiveSize = fixedLengthDeserializers.get(innerDeserializer.getClass());
+    }
+
+    @SuppressWarnings(value = "unchecked")
+    @Override
+    public void configure(Map<String, ?> configs, boolean isKey) {
+        if (inner == null) {
+            String listTypePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_TYPE_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_TYPE_CLASS;
+            String innerSerdePropertyName = isKey ? CommonClientConfigs.DEFAULT_LIST_KEY_SERDE_INNER_CLASS : CommonClientConfigs.DEFAULT_LIST_VALUE_SERDE_INNER_CLASS;
+            listClass = (Class<List<Inner>>) configs.get(listTypePropertyName);
+            Class<Serde> innerSerde = (Class<Serde>) configs.get(innerSerdePropertyName);
```

Comment:
nit: should be `Class<Deserializer<Inner>>` (2 times) -- (not `Serde` compare comment above) and we want to avoid warning about using a raw type

also `innerSerde -> innerDeserializerClass`